File: server/polar/void/meter/balance.py
import calendar
import math
from datetime import datetime, timedelta
from typing import Any, Literal

from pydantic import BaseModel, Field, ValidationError, computed_field

from polar.void.metric.schemas import TimeInterval

DELTAS = {
    TimeInterval.hour: timedelta(hours=1),
    TimeInterval.day: timedelta(days=1),
    TimeInterval.week: timedelta(weeks=1),
}


class Subscription(BaseModel):
    id: str
    at: datetime
    anchor: datetime
    meter_interval: TimeInterval
    meter_interval_count: int = Field(default=1, ge=1)
    rollover_cap: float | None = Field(default=None, ge=0)
    included: float = Field(default=0, ge=0, allow_inf_nan=False)
    limit: Literal["hard", "soft", "unlimited"] = "hard"
    ended: bool = False


class MeterEvent(BaseModel):
    id: str
    at: datetime
    name: str
    data: dict[str, Any]


class MeterCycle(BaseModel):
    subscription_id: str
    period_start: datetime
    period_end: datetime
    credits: float
    usage: float
    expired: float
    rollover: float
    overage: float


class Ledger(BaseModel):
    """What one identity holds on a meter, drained by its subtree's usage."""

    at: datetime | None = None
    subscription: Subscription | None = None
    boundary: datetime | None = None
    boundaries: list[datetime] = Field(default_factory=list)
    cycles: dict[datetime, MeterCycle] = Field(default_factory=dict)
    credits: float = 0
    usage: float = 0


class State(Ledger):
    """The ledger seen on its own: remaining and overage from its own credits."""

    @computed_field  # type: ignore[prop-decorator]
    @property
    def remaining(self) -> float:
        return max(self.credits - self.usage, 0)

    @computed_field  # type: ignore[prop-decorator]
    @property
    def overage(self) -> float:
        return max(self.usage - self.credits, 0)


def step(anchor: datetime, interval: TimeInterval, k: int) -> datetime:
    if interval in DELTAS:
        return anchor + DELTAS[interval] * k
    months = anchor.month - 1 + (k if interval == TimeInterval.month else 12 * k)
    year, month = anchor.year + months // 12, months % 12 + 1
    day = min(anchor.day, calendar.monthrange(year, month)[1])
    return anchor.replace(year=year, month=month, day=day)


def _meter_step(subscription: Subscription, k: int) -> datetime:
    return step(
        subscription.anchor,
        subscription.meter_interval,
        k * subscription.meter_interval_count,
    )


def _next_boundary(state: State) -> datetime | None:
    subscription = state.subscription
    if subscription is None or subscription.ended:
        return None
    floor = (
        max(subscription.at, state.boundary)
        if state.boundary is not None
        else subscription.at
    )
    if subscription.meter_interval in DELTAS:
        delta = DELTAS[subscription.meter_interval] * subscription.meter_interval_count
        k = max(0, math.ceil((floor - subscription.anchor) / delta))
    else:
        k = 0
        while _meter_step(subscription, k) < floor:
            k += 1
    boundary = _meter_step(subscription, k)
    if boundary == state.boundary:
        boundary = _meter_step(subscription, k + 1)
    return boundary


def _rollover(state: State) -> float:
    subscription = state.subscription
    assert subscription is not None
    rollover = state.remaining
    if subscription.rollover_cap is not None:
        rollover = min(rollover, subscription.rollover_cap)
    return rollover


def _advance(state: State, until: datetime) -> None:
    while (boundary := _next_boundary(state)) is not None and boundary <= until:
        assert state.subscription is not None
        if state.boundary is not None:
            rollover = _rollover(state)
            state.cycles[boundary] = MeterCycle(
                subscription_id=state.subscription.id,
                period_start=state.boundary,
                period_end=boundary,
                credits=state.credits,
                usage=state.usage,
                expired=state.remaining - rollover,
                rollover=rollover,
                overage=state.overage,
            )
            state.credits = rollover + state.subscription.included
            state.usage = 0
        state.boundary = boundary
        state.boundaries.append(boundary)


def _apply_stored_cycle(state: State, event: MeterEvent) -> None:
    if state.subscription is None or state.boundary != event.at:
        return
    try:
        stored = MeterCycle.model_validate(event.data)
    except ValidationError:
        return
    if stored.period_end != event.at:
        return
    if stored.subscription_id != state.subscription.id:
        return
    computed = state.cycles.get(event.at)
    if computed is None or computed.period_start != stored.period_start:
        return
    state.cycles[event.at] = stored
    state.credits = max(0, state.credits + stored.rollover - computed.rollover)


def _change(state: State, event: MeterEvent) -> None:
    subscription = state.subscription
    if event.name == "meter.cycled":
        _apply_stored_cycle(state, event)
        return
    if event.name == "subscription.cycled":
        return
    if event.name in ("subscription.canceled", "subscription.revoked"):
        if subscription is None:
            return
        subscription.ended = True
        subscription.at = event.at
        if event.name == "subscription.revoked":
            state.credits = 0
        return
    if event.name not in ("subscription.created", "subscription.updated"):
        return
    base = (
        subscription.model_dump()
        if subscription is not None and not subscription.ended
        else {"id": event.id, "anchor": event.at}
    )
    try:
        state.subscription = Subscription.model_validate(
            {**base, **event.data, "at": event.at, "ended": False}
        )
    except ValidationError:
        return
    previous = subscription.included if subscription and not subscription.ended else 0
    state.credits += max(0, state.subscription.included - previous)


def fold(
    state: State,
    events: list[MeterEvent],
    credits: list[tuple[datetime, float]],
    usage: list[tuple[datetime, float]],
    until: datetime,
) -> State:
    state = state.model_copy(deep=True)
    stream: list[tuple[datetime, int, Any]] = [
        (event.at, 0, event) for event in events if event.at <= until
    ]
    stream += [
        (timestamp, 1, value) for timestamp, value in credits if timestamp <= until
    ]
    stream += [(timestamp, 2, value) for timestamp, value in usage if timestamp < until]
    stream.sort(key=lambda item: item[:2])
    for timestamp, kind, item in stream:
        _advance(state, timestamp)
        if kind == 0:
            _change(state, item)
        elif kind == 1:
            state.credits = max(0, state.credits + item)
        else:
            state.usage += item
    _advance(state, until)
    state.at = until
    return state

File: clients/packages/void-sdk/src/storage/balance.ts
/** Mirrors server/void/meter/balance.py; timestamps are milliseconds internally. */
import type { MeterEvent, MeterCycle, Subscription } from '../api/generated'

export type Limit = 'hard' | 'soft' | 'unlimited'
interface Sub {
  id: string
  at: number
  anchor: number
  interval: Subscription['meter_interval']
  count: number
  cap: number | null
  /** Credits granted at the start of every period. */
  included: number
  limit: Limit
  ended: boolean
}
export interface FoldState {
  subscription: Sub | null
  boundary: number | null
  boundaries: number[]
  cycles: Map<number, MeterCycle>
  credits: number
  usage: number
}
export const emptyState = (): FoldState => ({
  subscription: null,
  boundary: null,
  boundaries: [],
  cycles: new Map(),
  credits: 0,
  usage: 0,
})
const iso = (at: number) => new Date(at).toISOString()
const remaining = (s: FoldState) => Math.max(s.credits - s.usage, 0)
const deltas: Partial<Record<Subscription['meter_interval'], number>> = {
  hour: 3600000,
  day: 86400000,
  week: 604800000,
}
function step(s: Sub, k: number): number {
  const delta = deltas[s.interval]
  if (delta !== undefined) return s.anchor + delta * k * s.count
  const anchor = new Date(s.anchor)
  const month =
    anchor.getUTCMonth() + k * s.count * (s.interval === 'year' ? 12 : 1)
  const date = new Date(s.anchor)
  date.setUTCDate(1)
  date.setUTCFullYear(anchor.getUTCFullYear(), month)
  const end = new Date(date)
  end.setUTCMonth(end.getUTCMonth() + 1, 0)
  date.setUTCDate(Math.min(anchor.getUTCDate(), end.getUTCDate()))
  return date.getTime()
}
/** The end of the current period, or null without a running subscription. */
export const nextBoundary = (s: FoldState): number | null => next(s)
function next(s: FoldState): number | null {
  const sub = s.subscription
  if (!sub || sub.ended) return null
  const floor = Math.max(sub.at, s.boundary ?? sub.at)
  const delta = deltas[sub.interval]
  let k = delta
    ? Math.max(0, Math.ceil((floor - sub.anchor) / (delta * sub.count)))
    : 0
  while (step(sub, k) < floor) k++
  let boundary = step(sub, k)
  if (boundary === s.boundary) boundary = step(sub, k + 1)
  return boundary
}
function advance(s: FoldState, until: number) {
  for (
    let boundary = next(s);
    boundary !== null && boundary <= until;
    boundary = next(s)
  ) {
    if (s.boundary !== null) {
      const rollover = Math.min(remaining(s), s.subscription!.cap ?? Infinity)
      s.cycles.set(boundary, {
        subscription_id: s.subscription!.id,
        period_start: iso(s.boundary),
        period_end: iso(boundary),
        credits: s.credits,
        usage: s.usage,
        expired: remaining(s) - rollover,
        rollover,
        overage: Math.max(s.usage - s.credits, 0),
      })
      s.credits = rollover + s.subscription!.included
      s.usage = 0
    }
    s.boundary = boundary
    s.boundaries.push(boundary)
  }
}
export function storedCycle(data: MeterEvent['data']): MeterCycle | null {
  const fields = ['credits', 'usage', 'expired', 'rollover', 'overage'] as const
  if (
    typeof data.subscription_id !== 'string' ||
    typeof data.period_start !== 'string' ||
    typeof data.period_end !== 'string'
  )
    return null
  if (
    ![data.period_start, data.period_end].every((t) =>
      Number.isFinite(Date.parse(t)),
    )
  )
    return null
  if (
    !fields.every(
      (f) => typeof data[f] === 'number' && Number.isFinite(data[f]),
    )
  )
    return null
  return {
    subscription_id: data.subscription_id,
    period_start: iso(Date.parse(data.period_start)),
    period_end: iso(Date.parse(data.period_end)),
    credits: Number(data.credits),
    usage: Number(data.usage),
    expired: Number(data.expired),
    rollover: Number(data.rollover),
    overage: Number(data.overage),
  }
}
function change(s: FoldState, event: MeterEvent) {
  const at = Date.parse(event.at),
    data = event.data
  if (event.name === 'meter.cycled') {
    const stored = storedCycle(data),
      computed = s.cycles.get(at)
    if (
      stored &&
      computed &&
      s.subscription &&
      s.boundary === at &&
      Date.parse(stored.period_end) === at &&
      stored.subscription_id === s.subscription.id &&
      stored.period_start === computed.period_start
    ) {
      s.cycles.set(at, stored)
      s.credits = Math.max(0, s.credits + stored.rollover - computed.rollover)
    }
    return
  }
  if (
    event.name === 'subscription.canceled' ||
    event.name === 'subscription.revoked'
  ) {
    if (!s.subscription) return
    s.subscription.ended = true
    s.subscription.at = at
    if (event.name === 'subscription.revoked') s.credits = 0
    return
  }
  if (!['subscription.created', 'subscription.updated'].includes(event.name))
    return
  const prior = s.subscription && !s.subscription.ended ? s.subscription : null
  const interval = data.meter_interval ?? prior?.interval
  const count = data.meter_interval_count ?? prior?.count ?? 1
  const cap =
    data.rollover_cap === undefined ? (prior?.cap ?? null) : data.rollover_cap
  const anchor =
    data.anchor === undefined
      ? (prior?.anchor ?? at)
      : Date.parse(String(data.anchor))
  const id = data.id ?? prior?.id ?? event.id
  const included = data.included ?? prior?.included ?? 0
  const limit = data.limit ?? prior?.limit ?? 'hard'
  if (
    !['hour', 'day', 'week', 'month', 'year'].includes(String(interval)) ||
    typeof count !== 'number' ||
    !Number.isInteger(count) ||
    count < 1 ||
    (cap !== null && (typeof cap !== 'number' || cap < 0)) ||
    typeof included !== 'number' ||
    !Number.isFinite(included) ||
    included < 0 ||
    !['hard', 'soft', 'unlimited'].includes(String(limit)) ||
    !Number.isFinite(anchor) ||
    typeof id !== 'string'
  )
    return
  s.subscription = {
    id,
    at,
    anchor,
    interval: interval as Sub['interval'],
    count,
    cap: cap as number | null,
    included,
    limit: limit as Limit,
    ended: false,
  }
  // The first period's allowance lands with the subscription; later ones at
  // each boundary in `advance`. An update grants only what the new terms add.
  s.credits += Math.max(0, included - (prior?.included ?? 0))
}
export function fold(
  initial: FoldState,
  events: readonly MeterEvent[],
  credits: readonly [number, number][],
  usage: readonly [number, number][],
  until: number,
): FoldState {
  const state = structuredClone(initial)
  const stream: {
    at: number
    kind: number
    event?: MeterEvent
    value?: number
  }[] = [
    ...events
      .filter((e) => Date.parse(e.at) <= until)
      .map((event) => ({ at: Date.parse(event.at), kind: 0, event })),
    ...credits
      .filter(([at]) => at <= until)
      .map(([at, value]) => ({ at, kind: 1, value })),
    ...usage
      .filter(([at]) => at < until)
      .map(([at, value]) => ({ at, kind: 2, value })),
  ]
  stream.sort((a, b) => a.at - b.at || a.kind - b.kind)
  for (const item of stream) {
    advance(state, item.at)
    if (item.event) change(state, item.event)
    else if (item.kind === 1)
      state.credits = Math.max(0, state.credits + item.value!)
    else state.usage += item.value!
  }
  advance(state, until)
  return state
}

/** Resume from a compact remote state, without loading completed cycles. */
export function resume(state: import('../api/generated').State): FoldState {
  const sub = state.subscription
  return {
    subscription: sub
      ? {
          id: sub.id,
          at: Date.parse(sub.at),
          anchor: Date.parse(sub.anchor),
          interval: sub.meter_interval,
          count: sub.meter_interval_count ?? 1,
          cap: sub.rollover_cap ?? null,
          included: sub.included ?? 0,
          limit: sub.limit ?? 'hard',
          ended: sub.ended ?? false,
        }
      : null,
    boundary: state.boundary ? Date.parse(state.boundary) : null,
    boundaries: [],
    cycles: new Map(),
    credits: state.credits ?? 0,
    usage: state.usage ?? 0,
  }
}
